Wrap and popup exceptions from Executable

mstv opened the pull request:
Fixes #7795

## Proposed changes

- wrap any exception from starting an external process into the new `ExternalOperationException`,
- show any `ExternalOperationException` to the user when it is created in order to avoid silent errors eaten by too generic `catch`
- let the user decide whether the `ExternalOperationException` shall be reported as a bug anyways if unhandled
- adapt the process creation and the exception handling in the `ScriptRunner`
- refactor and extend `MessageBoxes`

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![grafik](https://user-images.githubusercontent.com/36601201/85960429-9f33c780-b9a3-11ea-94fd-6a75ba5a7b8d.png)

### After

![grafik](https://user-images.githubusercontent.com/36601201/97794891-ec4cd200-1bff-11eb-8a98-9320c44a6a54.png)

![grafik](https://user-images.githubusercontent.com/36601201/97794954-bb20d180-1c00-11eb-9b61-dff7b6fa3deb.png)

## Test methodology <!-- How did you ensure quality? -->

- add NUnit tests
- manual with User Scripts

## Test environment(s) <!-- Remove any that don't apply -->
- Git Extensions 33.33.33
- Build f33a5b45015f282bbc024605917b72a37972b015
- Git 2.27.0.windows.1 (recommended: 2.28.0 or later)
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4250.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented:
This PR is marked as WIP/draft mainly because all uses of `Executable` and `Process.Start` must be changed to use the new `ExecutableFactory` yet.

So: Is the generic approach OK -- disregarding `ScriptRunner` (see below)?

The screenshots show an edge case: In case of the `ScriptRunner`, the `ExternalOperationException` rather should not popup early (with the misleading info / question regarding the bug report). Instead, it should be translated in order to show it with the context information, i.e. a single popup with:

```
Failed to execute script "test".
Reason: Failed to execute "cmdx".
Working directory: "F:\Build\gitextensions3_dev\"
Reason: The system cannot find the file specified
```

How to translate `ExternalOperationException`s without referencing UI code?

RussKie commented:
👀 but looks like a lot of code... 🤔

RussKie commented:
> * refactor and extend `MessageBoxes`

Could you please explain the reasoning for this change? 
Also it seems unrelated to the executable changes.

RussKie commented on GitCommands/Git/ExecutableException.cs:
```diff
@@ -0,0 +1,25 @@
+            : base(inner)
+        {
+            ProcessStartInfo = processStartInfo;
+            if (string.IsNullOrEmpty(ProcessStartInfo.WorkingDirectory))
+            {
+                ProcessStartInfo.WorkingDirectory = Directory.GetCurrentDirectory();
+            }
```
Why does the exception do this? The exception is merely a carrier of error information, it mustn't modify information it carries.

mstv replied:
> Why does the exception do this? The exception is merely a carrier of error information, it mustn't modify information it carries.

I agree in general. But in this case, it adds missing information.
Alternatively, `Executable`/`ProcessWrapper` could specify the `WorkingDirectory` always.

RussKie commented on GitCommands/Git/Executable.cs:
```diff
@@ -98,7 +128,15 @@ public ProcessWrapper(string fileName, string arguments, string workDir, bool cr
+                }
+                catch (Exception ex)
+                {
+                    _logOperation.LogProcessEnd(null);
```
What do you think about logging the exception here, so it can be viewed in the command log?
E.g.
```suggestion
                    _logOperation.LogProcessEnd(null, ex);
```

mstv replied:
> What do you think about logging the exception here, so it can be viewed in the command log?

👍 I will extend the classes `ProcessOperation` and CommandLogEntry.

RussKie commented on GitCommands/Git/Executable.cs:
```diff
@@ -31,21 +32,50 @@ public Executable([NotNull] Func<string> fileNameProvider, [NotNull] string work
+                    FileName = fileName,
+                    Arguments = args,
+                    WorkingDirectory = _workingDir
+                });
```
Why do we need all this plumbing? The exception is caught and then handled in `ProcessWrapper`?

mstv replied:
> Why do we need all this plumbing? The exception is caught and then handled in `ProcessWrapper`?

The exception is not consumed. It is wrapped into `ExecutableException`.

One could consider to replace the arguments `fileName, args, _workingDir` with a prefilled `ProcessStartInfo`.
Though the extension method `GetOutput` below would have to be adapted, too.

mstv commented:
I have completed this -- including proper handling for `ScriptRunner`.

RussKie commented:
I'll need some time for the review. There is a lot of new plumbing which makes me concerned and I need to some time to think through it.

RussKie commented:
👀

RussKie commented:
Thank you @mstv, I've reviewed the code but I can't stop thinking this adds a lot of ("enterprise level") complexity, and that we can achieve the desired with significantly less plumbing.
I've drafted a simpler alternative that I think should achieve the same but with less code. We will need to review an number of call sites that one way or another invoke external operations (e.g. git.exe, explorer.exe, etc.), but we'd have to review them with this change as well.

I'm happy to setup a call with you to talk through, if you like.

mstv commented:
I have removed `ExecutableException` and added `Dispose` to the `ProcessWrapper` ctor.

I agree that the TaskDialog will improve the MessageBox though with little more pressure against NBug reports as in #8358.

mstv commented:
Best reviewed commit by commit.

RussKie commented:
Thank you @mstv, but the static factories make me very uncomfortable. We are not using any other instances other than "Default", and this makes the factories look superficial.
That is, I can't see what we're gaining with such change:
```diff
-            _gitExecutable = new Executable(() => AppSettings.GitCommand, WorkingDir);
+            _gitExecutable = ExecutableFactory.Default.Create(() => AppSettings.GitCommand, WorkingDir);
```

The changes in MessageBoxes.cs feel like we're overcomplicating the implementation yet we're probably not delivering a good UX - i.e. if I read the code correctly there may be isntances where we're just showing an error with a "Ignore and continue" button. The value of such dialog is fairly low - if the user will ignore and continue anyway, why bother showing the dialog?

. . .

The changes in https://github.com/gitextensions/gitextensions/pull/8278/commits/2720f2fc4baa57e7160ad55d616705cb33d04040 modulo changes to OsShellUtil.cs are worth extracting and submitting as a separate PR, they are an improvement, and will make this work more targeted.

mstv commented:
Let's focus on *what* is to be changed first and _then_ think about the implementation and _how_ the SRP can be met.

> **Current behaviour**
#7714 wraps `Process.Start` in `try { } catch (Exception ex) { MessageBox.Show() }`.
**Expected behaviour**
We should replace all instances of `Process` with a generic wrapper -- at least all calls to `Process.Start`, e.g. with `GitCommands.Executable.ProcessWrapper`, which also needs generic error handling: https://github.com/gitextensions/gitextensions/issues/7574#issuecomment-588985889. 
Related to #6527.

> - wrap **any exception** from starting an **external** process into the new `ExternalOperationException`,
> - show any `ExternalOperationException` to the user __when it is created__ in order to __avoid silent errors__ eaten by too generic `catch`

Any external operation can fail. Usually, this is not caused by GE itself, i.e. it is no bug.
That's why I want to display the initial exception with context information to the user (unless the calling function is using exceptions for its normal operation, e.g. detection of the git executable).
This shall be done by one central wrapper, which is not overruled by too generic `try { } catch { // ignore }`. Silent errors are evil.

Running user scripts or shells is a special case (which I should not have chosen as the first example for the screenshots).
If the external processes fail this aint a bug of GE. That's why the error is shown to the user without the option to report a bug.

> we're just showing an error with a "Ignore and continue" button. The value of such dialog is fairly low - if the user will ignore and continue anyway, why bother showing the dialog?

Of course, we do not continue in the sense of just swallowing the exception but in "continuing to run GE".
So it is rather an "abort the current operation". You have chosen this wording instead of my "Do you want to cancel the operation and ignore the error?". I overtook it without rethinking.

> The changes in MessageBoxes.cs feel like we're overcomplicating the implementation

What do you think is unnecessary? Exceptions can be thrown in the background. So we must switch to the UI thread in order to display the TD. (In addition, `Form.ActiveForm` is `null` if the application is not the active application, e.g. if halt in a break point. I'd say this does not make much sense to me.)

> static factories make me very uncomfortable. We are not using any other instances other than "Default".

Last time, you complained about static classes. How to deal with `ExternalOperationExceptionFactory.OnException`? Just tell me. I think we need this indirection, else it is getting spaghetti.

ivangrek commented on GitCommands/Git/ExternalOperationExceptionFactory.cs:
```diff
@@ -0,0 +1,28 @@
+        /// </summary>
+        public static ExternalOperationExceptionFactory Default { get; } = new ExternalOperationExceptionFactory();
+
+        public event Action<ExternalOperationException, Handling> OnException;
```
Why do we need this event?

ivangrek commented on GitCommands/Git/ExternalOperationExceptionFactory.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public ExternalOperationException Create(string context, string command, string workingDirectory, Exception inner, Handling handling = Handling.OptionalBugReport)
+        {
+            ExternalOperationException ex = new ExternalOperationException(context, command, workingDirectory, inner);
+            OnException?.Invoke(ex, handling);
+            return ex;
```
Could we put this event in an exception class?
I see that creating an exception does not guarantee it will be thrown. At first glance, we might not be using a factory.

ivangrek commented on GitCommands/Git/ExecutableFactory.cs:
```diff
@@ -0,0 +1,26 @@
+        /// </summary>
+        public static ExecutableFactory Default { get; } = new ExecutableFactory();
+
+        public IExecutable Create([NotNull] Func<string> fileNameProvider,
+            [NotNull] string workingDir = "",
+            ExternalOperationExceptionFactory.Handling exceptionHandling = ExternalOperationExceptionFactory.Handling.OptionalBugReport,
+            [CanBeNull] string context = null)
+            => new Executable(fileNameProvider, workingDir, exceptionHandling, context);
+
+        public IExecutable Create([NotNull] string fileName,
+            [NotNull] string workingDir = "",
+            ExternalOperationExceptionFactory.Handling exceptionHandling = ExternalOperationExceptionFactory.Handling.OptionalBugReport,
+            [CanBeNull] string context = null)
+            => Create(() => fileName, workingDir, exceptionHandling, context);
```
I only see one place where this is used. 
Could you explain why we need these two methods and not one?

`Func<string> fileNameProvider` 

This is dependency, but it's not clear why.

ivangrek commented on GitCommands/Git/ExternalOperationException.cs:
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+
+namespace GitCommands
+{
+    /// <summary>
+    /// Exception class for failures of external operations,
+    /// e.g. file access or start of other executables, which always can fail due to removed or locked files.
+    /// </summary>
+    public class ExternalOperationException : Exception
```
Could we put this class in GitCommands.Exceptions or some other non-primary namespace?

ivangrek commented on GitUI/MessageBoxes.cs:
```diff
@@ -124,13 +134,116 @@ public static bool ConfirmUpdateSubmodules(IWin32Window win)
         public static void ShellNotFound([CanBeNull] IWin32Window owner)
             => ShowError(owner, Instance._shellNotFound.Text, Instance._shellNotFoundCaption.Text);
 
+        public static void Show([CanBeNull] IWin32Window owner, ExternalOperationException ex, ExternalOperationExceptionFactory.Handling exceptionHandling)
+        {
+            if (exceptionHandling == ExternalOperationExceptionFactory.Handling.None)
+            {
+                return;
+            }
+
```
Why do we need to use "IWin32Window owner".
Most likely this question applies to all methods in this class, but this one in particular. 
Are there any cases where the current window will not own the message box?

RussKie replied:
IIRC there are cases where owner is a different window.

ivangrek replied:
Do you know specific cases?

ivangrek commented on GitUI/OsShellUtil.cs:
```diff
@@ -1,29 +1,57 @@
-            Process.Start("explorer.exe", filePath);
+            try
+            {
+                ExecutableFactory.Default.Create("explorer.exe", exceptionHandling: ExternalOperationExceptionFactory.Handling.Show).Start(filePath);
```
Could we use a constant for the "explorer.exe" line?

ivangrek commented on GitUI/OsShellUtil.cs:
```diff
@@ -33,7 +61,14 @@ public static void OpenUrlInDefaultBrowser(string url)
         {
             if (!string.IsNullOrWhiteSpace(url))
             {
-                Process.Start(url);
```
Could we use a Systerm.Uri for url parameter?

ivangrek commented on GitCommands/Git/ExecutableFactory.cs:
```diff
@@ -22,5 +23,8 @@ public class ExecutableFactory
             ExternalOperationExceptionFactory.Handling exceptionHandling = ExternalOperationExceptionFactory.Handling.OptionalBugReport,
             [CanBeNull] string context = null)
             => Create(() => fileName, workingDir, exceptionHandling, context);
+
+        public IProcess Spawn(string arguments, [NotNull] string workingDir = "")
+            => Create(Application.ExecutablePath, workingDir, ExternalOperationExceptionFactory.Handling.OptionalBugReport).Start(arguments);
```
Why ExecutableFactory creates IProcess if IExecutable does?

ivangrek commented on GitCommands/Git/GitModule.cs:
```diff
@@ -51,7 +51,7 @@ public GitModule([CanBeNull] string workingDir)
             _indexLockManager = new IndexLockManager(this);
             _commitDataManager = new CommitDataManager(() => this);
-            _gitExecutable = new Executable(() => AppSettings.GitCommand, WorkingDir);
+            _gitExecutable = ExecutableFactory.Default.Create(() => AppSettings.GitCommand, WorkingDir);
```
Why do we need to use () => AppSettings.GitCommand not AppSettings.GitCommand?

RussKie commented:
> Let's focus on _what_ is to be changed first and _then_ think about the implementation and _how_ the SRP can be met.
> 
> Any external operation can fail. Usually, this is not caused by GE itself, i.e. it is no bug.
> That's why I want to display the initial exception with context information to the user (unless the calling function is using exceptions for its normal operation, e.g. detection of the git executable).
> This shall be done by one central wrapper, which is not overruled by too generic `try { } catch { // ignore }`. Silent errors are evil.

👍  


> Running user scripts or shells is a special case (which I should not have chosen as the first example for the screenshots).
> If the external processes fail this aint a bug of GE. That's why the error is shown to the user without the option to report a bug.

👍 


> > we're just showing an error with a "Ignore and continue" button. The value of such dialog is fairly low - if the user will ignore and continue anyway, why bother showing the dialog?
> 
> Of course, we do not continue in the sense of just swallowing the exception but in "continuing to run GE".
> So it is rather an "abort the current operation". You have chosen this wording instead of my "Do you want to cancel the operation and ignore the error?". I overtook it without rethinking.

This makes sense, now when you put it like this 👍 

I'm going to ignore the other two paragraphs and questions for the moment, and try to distill the requirements.

_In event when there is an unhandled exception (including `try... catch { /* do nothing */ }` unless documented why the exception is swallowed)_
_We want to display a meaningful user error including ability to report the issue to GitHub_
_Unless the error is caused by executing a user configurable functionality (such as scripts or shells) - in this case no GitHub reporting is provided_.

Did I get it correctly?

Currently all unhandled exceptions are trapped via `ReportBug(Exception)` method, that spawn the NBug dialog, if necessary.
Could this be the place to implement the logic that checks and show the error? E.g. something like this:
```cs
private static void ReportBug(Exception ex)
{
    if (ex is ExternalOperationException externalException)
    {
        // Something happened either in the app - can submit the bug to GitHub
        ReportAppException(externalException);
        return;
    }

    // Likely a user error
    ReportUserException(ex);
}
```

This way the implementation be simpler, in places where we know we can expect a failure we can `catch (Exception ex) { throw new ExternalOperationException(ex); }`, and it will fall through to `ReportBug` above.
With this I don't think we'd need factories or changes to messageboxes, as we'd implement a messagebox/taskdialog in `ReportAppException`.

Thoughts?

RussKie commented:
I have updated #8358 with the above proposal. I tried few different scenarios, like git.exe is missing, user script failed, a random failure... Seemed to achieve the desired.

![image](https://user-images.githubusercontent.com/4403806/99671179-c8293600-2ac5-11eb-90a6-513479e01ef0.png)
![image](https://user-images.githubusercontent.com/4403806/99671586-57cee480-2ac6-11eb-82cc-82728be713c3.png)

mstv commented:
> In event when there is an unhandled exception (including try... catch { /* do nothing */ } unless documented why the exception is swallowed)
We want to display a meaningful user error including ability to report the issue to GitHub
Unless the error is caused by executing a user configurable functionality (such as scripts or shells) - in this case no GitHub reporting is provided.

> Did I get it correctly?

Yes. But in contrast to your approach, I do comply to the rule "Never trust your clients!", i.e. in this case, I do not trust the GE implementation, where I have seen too many swallowed errors.

RussKie commented:
We shouldn't trust input data, but we should trust implementations we own ;)
We can/need to clean unnecessary catches.

On Sat, 21 Nov. 2020, 10:32 am Michael Seibt, <notifications@github.com>
wrote:

> In event when there is an unhandled exception (including try... catch { /*
> do nothing */ } unless documented why the exception is swallowed)
> We want to display a meaningful user error including ability to report the
> issue to GitHub
> Unless the error is caused by executing a user configurable functionality
> (such as scripts or shells) - in this case no GitHub reporting is provided.
>
> Did I get it correctly?
>
> Yes. But in contrast to your approach, I do comply to the rule "Never
> trust your clients!", i.e. in this case, I do not trust the GE
> implementation, where I have seen too many swallowed errors.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8278#issuecomment-731460100>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXQDPJLCNILDJSNOU6TSQ34ADANCNFSM4OKYBCGA>
> .
>

RussKie commented:
Closing this for now, since this has been largely affected by https://github.com/gitextensions/gitextensions/pull/8358.
